### PR TITLE
Task#7 rent

### PR DESCRIPTION
**Pull request by dimov05**

### CODE REVIEW

File: src/main/java/bg/libapp/libraryapp/event/BookAuditListener.java [hunk 1]
```diff
@@ -17,13 +17,27 @@ public BookAuditListener(BookAuditService bookAuditService) {
...
         bookAuditService.updateYearOfBook(event);
-
     }
+    @EventListener
+    public void handleUpdateQuantityOfBook(UpdateQuantityBookAuditEvent event) {
+        bookAuditService.updateQuantityOfBook(event);
+    }
+//    @EventListener
+//    public void handleUpdateQuantityOfBook(UpdateQuantityBookAuditEvent event) {
+//        bookAuditService.updateQuantityOfBook(event);
```
Comment (AtanasAtanasovv) [inline src/main/java/bg/libapp/libraryapp/event/BookAuditListener.java]: ? :D

File: src/main/java/bg/libapp/libraryapp/event/UpdateActiveStatusBookAuditEvent.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package bg.libapp.libraryapp.event;
+
+import bg.libapp.libraryapp.model.entity.Book;
+
+public class UpdateActiveStatusBookAuditEvent extends BaseBookAuditEvent {
+    private String fieldName;
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/event/UpdateActiveStatusBookAuditEvent.java]: extract the fields in another base class, that extends BaseBookAuditEvent and make all the statusBookAuditEvent classes to extend the new one

File: src/main/java/bg/libapp/libraryapp/model/dto/book/BookFilterRequest.java [hunk 3]
```diff
@@ -1,38 +1,49 @@
 package bg.libapp.libraryapp.model.dto.book;
 
+import jakarta.validation.constraints.Pattern;
 import org.hibernate.validator.constraints.Length;
 
 import java.util.Arrays;
 
 public class BookFilterRequest {
-//    @NotBlank(message = "Title should not be blank")
+    //    @NotBlank(message = "Title should not be blank")
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/model/dto/book/BookFilterRequest.java]: Review all the committed comments

File: src/main/java/bg/libapp/libraryapp/model/entity/Rent.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package bg.libapp.libraryapp.model.entity;
+
+import jakarta.persistence.*;
+
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/model/entity/Rent.java:11]: Check this IntelliJ setting for disabling wildcard imports - https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports

File: src/main/java/bg/libapp/libraryapp/service/BookService.java [hunk 5]
```diff
@@ -109,33 +112,40 @@ public BookDTO updateYear(String isbn, BookUpdateYearRequest bookUpdateYearReque
...
-                    return new BookNotFoundException(isbn);
-                });
+        Book bookToEdit = getBookByIsbnOrThrowException(isbn);
         String oldValuePublisher = bookToEdit.getPublisher();
         String newValuePublisher = bookUpdatePublisherRequest.getPublisher();
         if (!oldValuePublisher.equals(newValuePublisher)) {
             bookToEdit.setPublisher(bookUpdatePublisherRequest.getPublisher());
             bookRepository.saveAndFlush(bookToEdit);
-            eventPublisher.publishEvent(new UpdatePublisherBookAuditEvent(bookToEdit, oldValuePublisher));
+            eventPublisher.publishEvent(new UpdateYearBookAuditEvent(bookToEdit, oldValuePublisher));
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/service/BookService.java]: review the change of this method

File: src/main/java/bg/libapp/libraryapp/service/BookService.java [hunk 6]
```diff
@@ -220,4 +248,103 @@ private static Specification<Book> getBookSpecifications(BookFilterRequest bookF
...
+        String oldStatusString = String.valueOf(oldStatus);
+
+        if (newStatus && !oldStatus) {
+            setBookStatusAndReason(book, null, true);
+            eventPublisher.publishEvent(new UpdateActiveStatusBookAuditEvent(book, oldStatusString));
+            eventPublisher.publishEvent(new UpdateDeactivateReasonBookAuditEvent(book, oldReason));
+            logger.info("Activated book with isbn '" + book.getIsbn() + "' and params: " + book);
+        } else if (!newStatus) {
+            isValidDeactivateReason(newReason);
+            if (!oldStatus && !Objects.equals(oldReason, newReason)) {
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/service/BookService.java]: Probably better to use String equals to compare the values of both strings.

Change also the check and extract the !oldStatus as we discussed

File: src/main/java/bg/libapp/libraryapp/service/RentService.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    private User userEligibleToRent(RentAddRequest rentAddRequest) {
+        String username = SecurityContextHolder.getContext().getAuthentication().getName();
+        logger.info("userEligibleToRent method called with params: " + rentAddRequest + "\n" +
+                " by authenticated user with username '" + username + "'.");
+        User borrower = userService.getUserByUsername(username);
+        String userRole = Role.values()[borrower.getRole()].toString();
+        Long userIdToRent = rentAddRequest.getUserId();
+        if (!userRole.equals("ADMIN")
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/service/RentService.java]: Use the enum values for the roles (or at least extract the string values in constants)

File: src/main/java/bg/libapp/libraryapp/service/RentService.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+                hasUserProlongedRents(borrower, rent);
+                if (hasEnoughTimeToReturnBook(rent)) {
+                    counterOfCurrentRents = countBorrowedBookAndThrowExceptionWhenMoreThanTwo(borrower, counterOfCurrentRents);
+                }
+            }
+        }
+    }
+
+    private void isBookCurrentlyRentedByUser(String isbn, Rent rent) {
+        if (rent.getBook().getIsbn().equals(isbn) && rent.getActualReturnDate() != null) {
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/service/RentService.java]: reverse the check for the actualReturnDate (as you already found the problem, this comment is just for a reminder :) )

File: src/main/java/bg/libapp/libraryapp/service/BookService.java [hunk 9]
```diff
@@ -119,14 +126,21 @@ private void isBookActive(Book bookToEdit) {
         }
     }
 
+    private void isBookInActive(Book bookToEdit) {
+        if (bookToEdit.isActive()) {
+            logger.error("Book with isbn '" + bookToEdit.getIsbn() + "' is not active!");
```
Comment (KrisDoseva) [inline src/main/java/bg/libapp/libraryapp/service/BookService.java:131]: Here the check in the IF is whether the book IS active and the error is that it is not, please, double check this

**Review by KrisDoseva — APPROVED** [review]

**Pull request merged**